### PR TITLE
[servus] extractor reimplementation due to new url schema

### LINKED ISSUE #1490: ServusTV not working
### Checklist

- [X] I'm reporting a broken site
- [X] I've verified that I'm running yt-dlp version **2021.10.22**. ([update instructions](https://github.com/yt-dlp/yt-dlp#update))
- [X] I've checked that all provided URLs are alive and playable in a browser
- [X] I've checked that all URLs and arguments with special characters are [properly quoted or escaped](https://github.com/ytdl-org/youtube-dl#video-url-contains-an-ampersand-and-im-getting-some-strange-output-1-2839-or-v-is-not-recognized-as-an-internal-or-external-command)
- [X] I've searched the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues including closed ones. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [ ] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and I'm willing to share it if required

### Region

Austria/Europe

### Description

Hey, guys!

According to --list-extractors, Servus is included in the list, but does unfortunately not work properly.
I've been trying to download this documentary:
https://www.servustv.com/volkskultur/v/aa-28jq9b51h1w11/
yt-dlp does download and convert an audio-stream, but **no video**, and then it puts it into a .mp4 file-format all the same.
**VLC**, however, does the exact opposite; it opens a video-stream, but with **no audio**, when I enter the very same link, that I have extracted from the site, and that I also used for yt-dlp.
The latter is pretty odd, since the link contains a folder "audio".
https://dms.redbull.tv/v4/dms/resource/stv/audio/fmp4/AA-28JQ9B51H1W11/de/personal_computer/chrome/at/playlist.m3u8
So, obviously a link to a play-list.

Depending on what download-helper I use with yt-dlp, I do or don't get to see all play-list items during download, which are around a thousand for an end-product of a 53,7MB sized audio-file, but named Play-List.mp4.

Since VLC does get a video stream under this link, both formats, audio and video, should actually be available, but somehow, they don't seem to be accessible to yt-dlp.

Since you're the pros, you will certainly figure it out.
And since this site does broadcast pretty decent documentaries, I'd appreciate it very much, if you'd undergo the effort to fix it!

Thanks in advance, and yours, azteca

### Verbose log

```shell
[debug] Command-line config: ['-Uv']
[debug] Encodings: locale UTF-8, fs utf-8, out UTF-8, err UTF-8, pref UTF-8[debug] yt-dlp version 2021.10.22 (zip)
[debug] Python version 3.7.3 (CPython 64bit) - Linux-4.19.0-17-rt-amd64-x86_64-with-debian-10.11
[debug] exe versions: ffmpeg 4.1.8-0, ffprobe 4.1.8-0, phantomjs 2.1.1
[debug] Optional libraries: Cryptodome, mutagen, sqlite
[debug] Proxy map: {}
yt-dlp is up to date (2021.10.22)
```

### PR DESCRIPTION
**Pull request by flashdagger**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information
Closes #1490
Current servustv.com platform seems to have an API which is easy to implement.
Since I don't know the old one I left the ServusIE class untouched.

Open questions:
- video and audio streams are separated - can we provide a default merge?

### CODE REVIEW

**Comment by pukkandan** [comment]
> can we provide a default merge?

wdym? We already merge the best video and audio together assuming ffmpeg is available

**Comment by flashdagger** [comment]
> > can we provide a default merge?
> 
> wdym? We already merge the best video and audio together assuming ffmpeg is available

Just realized it myself. Very smart! But for the test I had to specify a format, otherwise it says 'url field is missing' which made me puzzled.

**Comment by pukkandan** [comment]
for the test, specify `bestvideo` as format

**Comment by flashdagger** [comment]
Another thing. The JSON API requires a timezone in the query. Something like `Europe/Berlin`.  It also seems to work with invalid stuff line `XX`. You guys have any experience with that?

**Comment by pukkandan** [comment]
Only site that I know needs it is tiktok. There, we have just hardcoded it to New York. Assuming it doesn't actually affect the extraction, ig you can similarly hardcode any timezone

File: yt_dlp/extractor/servus.py [hunk 1]
```diff
@@ -1,148 +1,304 @@
...
+            raise ExtractorError(
+                errormsg, video_id=video_id, expected=True)
+
+        try:
+            formats, subtitles = self._extract_m3u8_formats_and_subtitles(
+                info['videoUrl'], video_id=video_id,
+                entry_protocol='m3u8' if is_live else 'm3u8_native',
+                errnote='Stream not available')
+        except ExtractorError as exc:
+            raise ExtractorError(exc.msg, video_id=video_id, expected=True)
```
Comment (pukkandan) [inline yt_dlp/extractor/servus.py:111]:
```suggestion
        formats, subtitles = self._extract_m3u8_formats_and_subtitles(
            info['videoUrl'], video_id, 'mp4', live=is_live, errnote=False)
        if not formats:
            raise ExtractorError('Stream not available', video_id=video_id, expected=True)
```

File: yt_dlp/extractor/servus.py [hunk 2]
```diff
@@ -1,148 +1,304 @@
...
-        series = attrs.get('label')
-        season = attrs.get('season')
-        episode = attrs.get('chapter')
-        duration = float_or_none(attrs.get('duration'), scale=1000)
-        season_number = int_or_none(self._search_regex(
-            r'Season (\d+)', season or '', 'season number', default=None))
-        episode_number = int_or_none(self._search_regex(
-            r'Episode (\d+)', episode or '', 'episode number', default=None))
+        info.setdefault('videoUrl', video_url)
+        errors = ", ".join(info.get('playabilityErrors', ()))
```
Comment (pukkandan) [inline yt_dlp/extractor/servus.py:96]: use single quotes everywhere

File: yt_dlp/extractor/servus.py [hunk 3]
```diff
@@ -1,148 +1,304 @@
...
+                    url, ie=ie or self.ie_key(), video_id=video_id, video_title=title)
+
+        return OnDemandPagedList(fetch_page, page_size)
+
+    @staticmethod
+    def _page_id(json_obj):
+        for key, value in traverse_obj(json_obj, ('source', 'data'), default={}).items():
+            if isinstance(value, dict) and 'id' in value:
+                return value['id']
+        return None
```
Comment (pukkandan) [inline yt_dlp/extractor/servus.py:170]: this shouldnt even be a function
```suggestion
    def _page_id(json_obj):
        return traverse_obj(json_obj, ('source', 'data', ..., 'id'))
```

File: yt_dlp/extractor/servus.py [hunk 4]
```diff
@@ -1,148 +1,304 @@
...
                     '''
+    _GEO_COUNTRIES = ['AT', 'DE', 'CH', 'LI', 'LU', 'IT']
+    _GEO_BYPASS = False
+    _API_URL = 'https://api-player.redbull.com/stv/servus-tv'
+    _QUERY_API_URL = 'https://backend.servustv.com/wp-json/rbmh/v2/query-filters/query/'
+    _LIVE_URLS = {
+        'AT': 'https://dms.redbull.tv/v4/destination/stv/stv-linear'
+              '/personal_computer/chrome/at/de_AT/playlist.m3u8',
+        'DE': 'https://dms.redbull.tv/v4/destination/stv/stv-linear'
+              '/personal_computer/chrome/de/de_DE/playlist.m3u8',
```
Comment (pukkandan) [inline yt_dlp/extractor/servus.py:39]: read contributing guidelines. It specifically says DO NOT split literals like this
```suggestion
        'AT': 'https://dms.redbull.tv/v4/destination/stv/stv-linear/personal_computer/chrome/at/de_AT/playlist.m3u8',
        'DE': 'https://dms.redbull.tv/v4/destination/stv/stv-linear/personal_computer/chrome/de/de_DE/playlist.m3u8',
```

also, it appears only the language code in the URL is different. In that case, a mapping like this is unnecessary

**Review by pukkandan — CHANGES_REQUESTED** [review]
PS: the code is all over the place and really hurts readability. if possible, try to reorganize it to have a more linear flow

**Comment by flashdagger** [comment]
implemented as separate plugin

**Comment by pukkandan** [comment]
Making me review the PR and then closing it is not appreciated. If you don't intend to address reviews, don't make a PR in the first place

**Comment by MuenchenSued** [comment]
> implemented as separate plugin

Hello flashdagger, I feel disappointed, that you've decided to fork this project and implement a solution for yourself instead of contributing here with your knowledge. A couple of people are waiting for a solution and it is no good idea to install a separate tool or plugin for each website. I would appreciate if you could adapt your solution according pukkandan recommendations to make yt-dlp better and better.

**Comment by flashdagger** [comment]
> > implemented as separate plugin
> 
> Hello flashdagger, I feel disappointed, that you've decided to fork this project and implement a solution for yourself instead of contributing here with your knowledge. A couple of people are waiting for a solution and it is no good idea to install a separate tool or plugin for each website. I would appreciate if you could adapt your solution according pukkandan's recommendations to make yt-dlp better and better.

A plugin is not a fork. You just extend the official yt-dlp installation with additional extractors and still benefit from all improvements of the core library. That way other developers can provide extractors which the yt-dlp team doesn't have to maintain.

**Comment by chrizilla** [comment]
@flashdagger: You seem to have a working extractor for servustv.com. Could you offer your code to the yt-dlp team?

**Comment by flashdagger** [comment]
> @flashdagger: You seem to have a working extractor for servustv.com. Would you be willing to share your code with the yt-dlp team?

Sure. It's Unlicense anyway.

**Comment by chrizilla** [comment]
ok, sry I wasn't clear, I meant giving pukkandan a link to your code.

**Pull request closed without merging**